### PR TITLE
Edge-router/ingress-controller support.

### PR DESCRIPTION
**Pull request by enxebre**

### CODE REVIEW

File: roles/addons/files/traefik.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+defaultEntryPoints = ["http"]
+[entryPoints]
+  [entryPoints.http]
+  address = ":80"
```
Comment (enxebre) [inline roles/addons/files/traefik.toml]: will create follow up for https/lets encrypt here

File: terraform/aws/public-cloud/edge-routers.tf [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    etcd_ca            = "${replace(module.ca.ca_cert_pem, \"\n\", \"\\n\")}"
+    etcd_cert          = "${replace(module.etcd_cert.etcd_cert_pem, \"\n\", \"\\n\")}"
+    etcd_key           = "${replace(module.etcd_cert.etcd_private_key, \"\n\", \"\\n\")}"
+  }
+}
+
+resource "aws_instance" "edge-router" {
+  instance_type        = "${var.edge-router_instance_type}"
+  ami                  = "${module.edge-router_ami.ami_id}"
+  // iam_instance_profile = "${module.iam.edge-router_profile_name}"
```
Comment (tayzlor) [inline terraform/aws/public-cloud/edge-routers.tf]: this needs resolved

File: roles/addons/tasks/traefik.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  tags:
+    - addons
+    - traefik-ingress-controller
+
+- name: create traefik config dir
+  file:
+    path: "/etc/traefik"
+    state: directory
+    mode: 0755
+  delegate_to: "{{item}}"
```
Comment (tayzlor) [inline roles/addons/tasks/traefik.yml]: formatting issues (spaces around vars in some places others not)

File: roles/addons/tasks/traefik.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    namespace: kube-system
+    filename: "{{ kubernetes_addons_dir }}/traefik-ingress-controller.yaml"
+    state: present
+  tags:
+    - addons
+    - traefik-ingress-controller
+
+- name: create traefik config dir
+  file:
+    path: "/etc/traefik"
```
Comment (tayzlor) [inline roles/addons/tasks/traefik.yml]: dont need quotes here

**Comment by tayzlor** [comment]
think this needs a follow up to investigate ASGs on the workers and edge routers

File: group_vars/all.yml [hunk 5]
```diff
@@ -82,3 +82,8 @@ dns_domain: cluster.local
 
 # Toggle for cloud provider support for Kubelet
 enable_cloud_provider: false
+
+# Choose and specific ingress-controller. Supported:
```
Comment (tayzlor) [inline group_vars/all.yml]: grammar here - Choose and?

File: roles/addons/files/traefik-ingress-controller.yaml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          hostPort: 8080
+        volumeMounts:
+          - mountPath: /etc/traefik
+            name: test-volume
+            readOnly: false
+        args:
+        - --web
+        - --kubernetes
+        - --configFile=/etc/traefik/traefik.toml
+        - --logLevel=DEBUG
```
Comment (tayzlor) [inline roles/addons/files/traefik-ingress-controller.yaml]: should the log level be configurable here?

File: roles/addons/files/traefik-ingress-controller.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      volumes:
+        - hostPath:
+            path: /etc/traefik
+          name: test-volume
```
Comment (tayzlor) [inline roles/addons/files/traefik-ingress-controller.yaml]: test-volume? prob warrants a better name?

File: roles/kube-worker/templates/kubelet.service.j2 [hunk 8]
```diff
@@ -11,6 +11,8 @@ Environment=KUBELET_VERSION={{ kubernetes_version }}
 ExecStart=/opt/bin/kubelet-wrapper \
   --api-servers={{ kube_api_servers }} \
   --allow-privileged=true \
+  --register-schedulable={{ schedulable }} \
+  --node-labels="{{ node_labels }}" \
```
Comment (tayzlor) [inline roles/kube-worker/templates/kubelet.service.j2]: seems a bit odd to set ```--node-labels=""``` as a default - should we be a bit more defensive here?

File: terraform/aws/public-cloud/main.tf [hunk 9]
```diff
@@ -113,9 +116,9 @@ module "docker_client_certs" {
   source                = "github.com/Capgemini/tf_tls/docker/client"
   ca_cert_pem           = "${module.ca.ca_cert_pem}"
   ca_private_key_pem    = "${module.ca.ca_private_key_pem}"
-  ip_addresses_list     = "${concat(aws_instance.master.*.private_ip, aws_instance.worker.*.private_ip)}"
-  deploy_ssh_hosts      = "${concat(aws_instance.master.*.public_ip, aws_instance.worker.*.public_ip)}"
-  docker_client_count   = "${var.masters + var.workers}"
+  ip_addresses_list     = "${concat(aws_instance.edge-router.*.private_ip, concat(aws_instance.master.*.private_ip, aws_instance.worker.*.private_ip))}"
```
Comment (tayzlor) [inline terraform/aws/public-cloud/main.tf]: is the nested ```concat``` deliberate here (and in other instances)?

Comment (enxebre) [inline terraform/aws/public-cloud/main.tf]: Yeah, that's the quickest way I found to concat three of them masters, workers and edge-routers

**Pull request merged**